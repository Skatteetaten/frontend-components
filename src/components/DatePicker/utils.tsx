export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
