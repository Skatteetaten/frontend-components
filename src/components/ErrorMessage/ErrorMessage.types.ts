export interface ErrorMessageProps {
  /** Overstyring av stiler */
  className?: string;
  /** Dersom feilmeldingen skal vises eller ikke */
  showError?: boolean;
  /** Feilmelding */
  children: JSX.Element | string;
}
