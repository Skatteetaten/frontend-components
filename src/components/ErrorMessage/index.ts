export { default } from './ErrorMessage';
export * from './ErrorMessage';
