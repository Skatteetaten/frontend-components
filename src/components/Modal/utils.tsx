import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
