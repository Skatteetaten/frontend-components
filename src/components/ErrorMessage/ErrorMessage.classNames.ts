import { mergeStyles } from '@uifabric/merge-styles';
import {
  getTheme,
  AnimationClassNames,
  getGlobalClassNames
} from '@uifabric/styling';
import { MdIcons } from '../utils/icons/';
import { PaletteProps } from '..';

const errorIcon = "'" + MdIcons.icons.Error + "'";
export var getClassNames = function getClassNames() {
  const theme = getTheme();
  const { fonts } = theme;
  const palette = theme.palette as PaletteProps;
  const classNames = getGlobalClassNames({}, theme);

  return mergeStyles([
    // @ts-ignore TODO
    classNames.error,
    AnimationClassNames.slideDownIn20,
    fonts.small,
    {
      display: 'block',
      color: palette.skeColor.error,
      fontWeight: 500,
      paddingTop: 5,
      selectors: {
        ':before': {
          fontFamily: MdIcons.fontFace.fontFamily,
          fontSize: fonts.small,
          content: errorIcon,
          marginRight: 3
        }
      }
    }
  ]);
};
