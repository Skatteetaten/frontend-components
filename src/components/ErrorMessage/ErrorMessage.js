import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@uifabric/utilities';
import { getClassNames } from './ErrorMessage.classNames';

const ErrorMessage = props => {
  const errorClassNames = getClassNames(props);
  const showError = props.showError !== undefined ? props.showError : true;
  if (!showError) return null;
  return (
    <span role="alert" className={css(errorClassNames)}>
      {props.children}
    </span>
  );
};

export default ErrorMessage;

ErrorMessage.propTypes = {
  children: PropTypes.any,
  showError: PropTypes.bool
};
