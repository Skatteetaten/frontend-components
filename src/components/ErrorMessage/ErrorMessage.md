** ErrorMessage brukes til å vise feilmeldinger. **

```js
import ErrorMessage from '@skatteetaten/frontend-components/ErrorMessage';

<ErrorMessage>Skriv datoen slik: 17.05.2019</ErrorMessage>;
```

Feilmelding ved en komponent:

```js
import ErrorMessage from '@skatteetaten/frontend-components/ErrorMessage';
import Button from '@skatteetaten/frontend-components/Button';
const initialState = {
  hasError: false
};

<>
  <Button
    type="primary"
    onClick={() => setState({ hasError: !state.hasError })}
  >
    {state.hasError ? 'Skjul feil' : 'Vis Feil'}
  </Button>
  <ErrorMessage showError={state.hasError}>
    Du har 3 valideringsfeil.
  </ErrorMessage>
</>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      Dersom komponentene ikke har noen innebygget mekanisme for feilmeldinger,
      kan ErrorMessage brukes for å vise feilmeldinger knyttet til et mindre
      element. Typiske situasjoner er felt eller knapper i skjemaer.
      Feilmeldingen skal vises i umiddelbar nærhet til elementet den hører til.
    </p>
    <p>
      Teksten i feilmeldingen bør hjelpe brukeren å komme videre, fremfor å peke
      på hva han/hun har gjort feil. Du kan for eksempel skrive: "Skriv datoen
      slik: 17.05.2019" i stedet for "Du har brukt feil datoformat."
    </p>
    <p>For feil eller varsler som gjelder hele siden, bruk MessageBar.</p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <p>
      Sjekk at den røde teksten plasseres på en flate som gir tilstrekkelig god
      lesekontrast.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      Dette er en egenutviklet komponent, så ingen flere props er tilgjengelig.
    </p>
  </AccordionItem>
</Accordion>;
```
