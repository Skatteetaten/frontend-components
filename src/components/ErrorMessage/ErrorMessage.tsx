import { css } from '@uifabric/utilities';
import * as React from 'react';
import { getClassNames } from './ErrorMessage.classNames';

interface ErrorMessageProps {
  children: JSX.Element | string;
  showError?: boolean;
}

const ErrorMessage = (props: ErrorMessageProps) => {
  // @ts-ignore TODO
  const errorClassNames = getClassNames(props);
  const showError = props.showError !== undefined ? props.showError : true;
  if (!showError) {
    return null;
  }
  return (
    <span role="alert" className={css(errorClassNames)}>
      {props.children}
    </span>
  );
};

export default ErrorMessage;
