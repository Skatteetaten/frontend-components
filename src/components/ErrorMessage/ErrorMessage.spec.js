import React from 'react';
import { shallow } from 'enzyme';
import ErrorMessage from './ErrorMessage';
import { matches } from '../utils/test-utils';

function oppsettShallow(props) {
  return shallow(<ErrorMessage {...props} />);
}

describe('ErrorMessage komponent', () => {
  it('matcher snapshot', () => {
    matches(<ErrorMessage> FeilMelding </ErrorMessage>);
  });

  it('Rendrer uten feil', () => {
    const wrapper = oppsettShallow({});
    expect(wrapper.exists()).toEqual(true);
  });

  it('Skjuler komponent hvis showError er false', () => {
    const wrapper = oppsettShallow({ showError: false });
    matches(wrapper);
    expect(wrapper.exists()).toEqual(true);
  });
});
