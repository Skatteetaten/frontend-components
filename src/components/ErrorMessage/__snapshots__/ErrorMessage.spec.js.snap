// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage komponent Skjuler komponent hvis showError er false 1`] = `null`;

exports[`ErrorMessage komponent matcher snapshot 1`] = `
<span
  aria-atomic="true"
  className="css-34"
  role="alert"
>
   FeilMelding 
</span>
`;
